### Student Information
Name: 余承祐

Student ID: 111062233

GitHub ID: OmeiLab

---

### Instructions

### First Phase Submission

1. First: do the **take home** exercises in the [DM2025-Lab1-Master](https://github.com/leoson-wu/DM2025-Lab1-Exercise/blob/main/DM2025-Lab1-Master.ipynb) that considered as **phase 1 (from exercise 1 to exercise 15)**. You can answer in the master file. __This part is worth 10% of your grade.__


2. Second: follow the same process from the [DM2025-Lab1-Master](https://github.com/leoson-wu/DM2025-Lab1-Exercise/blob/main/DM2025-Lab1-Master.ipynb) on **the new dataset** up **until phase 1**. You can skip some exercises if you think some steps are not necessary. However main exercises should be completed. You don't need to explain all details as we did (some **minimal comments** explaining your code are useful though).  __This part is worth 15% of your grade.__
    -  Use [the new dataset](https://github.com/leoson-wu/DM2025-Lab1-Exercise/blob/main/newdataset/Reddit-stock-sentiment.csv). The dataset contains a 16 columns including 'text' and 'label', with the sentiment labels being: 1.0 is positive, 0.0 is neutral and -1.0 is negative. You can simplify the dataset and use only the columns that you think are necessary. 
    
    - You are allowed to use and modify the `helper` functions in the folder of the first lab session (notice they may need modification) or create your own.
    - Use this file to complete the homework from the second part. Make sure the code can be run from the beginning till the end and has all the needed output.


3. Third: please attempt the following tasks on **the new dataset**. __This part is worth 10% of your grade.__
    - Generate meaningful **new data visualizations**. Refer to online resources and the Data Mining textbook for inspiration and ideas. 
    


4. Fourth: It's hard for us to follow if your code is messy, so please **tidy up your notebook** and **add minimal comments where needed**. __This part is worth 5% of your grade.__

You can submit your homework following these guidelines: [DM2025-Lab1-announcement](https://github.com/leoson-wu/DM2025-Lab1-Announcement/blob/main/README.md). Make sure to commit and save your changes to your repository __BEFORE the deadline (September 28th 11:59 pm, Sunday)__. 

### Second Phase Submission 

**You can keep the answer for phase 1 for easier running and update the phase 2 on the same page.**

1. First: Continue doing the **take home** exercises in the [DM2025-Lab1-Master](https://github.com/leoson-wu/DM2025-Lab1-Exercise/blob/main/DM2025-Lab1-Master.ipynb) for **phase 2, starting from Finding frequent patterns**. Use the same master(.ipynb) file. Answer from phase 1 will not be considered at this stage. You can answer in the master file. __This part is worth 10% of your grade.__


2. Second: Continue from first phase and do the same process from the [DM2025-Lab1-Master](https://github.com/leoson-wu/DM2025-Lab1-Exercise/blob/main/DM2025-Lab1-Master.ipynb) on **the new dataset** for phase 2, starting from Finding frequent pattern. You can skip some exercises if you think some steps are not necessary. However main exercises should be completed. You don't need to explain all details as we did (some **minimal comments** explaining your code are useful though).  __This part is worth 15% of your grade.__
    - Continue using this file to complete the homework from the second part. Make sure the code can be run from the beginning till the end and has all the needed output. Use the same new dataset as in phase 1.
    
    - You are allowed to use and modify the `helper` functions in the folder of the first lab session (notice they may need modification) or create your own.

3. Third: please attempt the following tasks on **the new dataset**. __This part is worth 20% of your grade.__
    - Use this file to answer.
    - Generate **TF-IDF features** from the tokens of each text. This will generating a document matrix, however, the weights will be computed differently (using the TF-IDF value of each word per document as opposed to the word frequency).  Refer to this Scikit-learn [guide](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) .
    - Implement a simple **Naive Bayes classifier** that automatically classifies the records into their categories. Use both the TF-IDF features and word frequency features to build two seperate classifiers. Note that for the TF-IDF features you might need to use other type of NB classifier different than the one in the Master Notebook. Comment on the differences and when using augmentation with feature pattern.  Refer to this [article](https://hub.packtpub.com/implementing-3-naive-bayes-classifiers-in-scikit-learn/).


4. Fourth: In the lab, we applied each step really quickly just to illustrate how to work with your dataset. There are somethings that are not ideal or the most efficient/meaningful. Each dataset can be handled differently as well. What are those inefficent parts you noticed? How can you improve the Data preprocessing for these specific datasets? __This part is worth 10% of your grade.__


5. Fifth: It's hard for us to follow if your code is messy, so please **tidy up your notebook** and **add minimal comments where needed**. __This part is worth 5% of your grade.__


You can submit your homework following these guidelines: [DM2025-Lab1-announcement](https://github.com/leoson-wu/DM2025-Lab1-Announcement/blob/main/README.md). Make sure to commit and save your changes to your repository __BEFORE the deadline (October 19th 11:59 pm, Sunday)__. 

# Phase 1

In [9]:
### Begin Assignment Here
# import necessary dataset
import pandas as pd
import numpy as np
import nltk
from sklearn.feature_extraction.text import CountVectorizer
import matplotlib.pyplot as plt
import seaborn as sns
import plotly as py

nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\user\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\user\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

## Data Preparation

In [4]:
df = pd.read_csv('newdataset/Reddit-stock-sentiment.csv')
df.head()	# check the top 5 rows of the dataset

type             datetime  post_id         subreddit  \
0  comment  2025-04-11 17:29:56  mmli62w    wallstreetbets   
1  comment   2025-04-12 1:12:19  mmnu7v9    wallstreetbets   
2  comment  2025-04-10 15:09:41  mmeevio       StockMarket   
3     post  2023-08-30 17:12:55  165kllm  stockstobuytoday   
4  comment  2025-04-11 14:48:05  mmkl6bw       StockMarket   

                                               title                author  \
0    Retardation is on the menu boys! WSB is so back          StickyTip420   
1  Retail giant TARGET has now declined for 10 co...  Comfortable-Dog-8437   
2  How do you feel about a sitting president maki...          Btankersly66   
3                              Who knows more? $VMAR             emiljenfn   
4  The Trump administration is begging Xi Jinping...          Just-Big6411   

                                                 url  upvotes  downvotes  \
0               https://i.redd.it/0yq2ftren8ue1.jpeg        0        NaN   
1               https://i.redd.it/7tl6puv9waue1.jpeg      -15        NaN   
2  https://apnews.com/article/trump-truth-social-...        1        NaN   
3  https://www.reddit.com/r/stockstobuytoday/comm...       30        0.0   
4  https://edition.cnn.com/2025/04/10/politics/tr...        1        NaN   

   upvote_ratio                                               text  \
0           NaN                                   Calls on retards   
1           NaN  Stunt as in like why did they even make a big ...   
2           NaN                  Seeing lots of red in the ticker.   
3          0.98  Vision Marine Technologies Inc. is rewriting t...   
4           NaN                           He didn’t say thank you.   

   subjectivity  polarity  sentiment  \
0      1.000000 -0.900000       -1.0   
1      0.177778  0.083333        1.0   
2      0.000000  0.000000        0.0   
3      0.646970  0.216383        1.0   
4      0.000000  0.000000        0.0   

                                            entities  label  
0                                                 []   -1.0  
1              ['Stunt', 'company', 'deal', 'place']    0.0  
2                                         ['ticker']    0.0  
3  ['watercraft', 'skill', 'power', ']', 'feat', ...    1.0  
4                                                 []   -1.0

Let's generate a new dataset by removing some redundant columns. We will preserve these:
- `type`
  
- `subreddit`
- `title`
- `upvotes`
- `downvotes`
- `upvote_ratio`
- `text`
- `subjectivity`
- `polarity`
- `sentiment`
- `entities`
- `label`

These columns are just chosen by my intuition, that is, some columns will not be used then or even be dropped.

In [13]:
data = df[['type', 'subreddit', 'title', 'upvotes', 'downvotes', 'upvote_ratio', 'text', 'subjectivity', 'polarity', 'sentiment', 'entities', 'label']]
data.head()

type         subreddit  \
0  comment    wallstreetbets   
1  comment    wallstreetbets   
2  comment       StockMarket   
3     post  stockstobuytoday   
4  comment       StockMarket   

                                               title  upvotes  downvotes  \
0    Retardation is on the menu boys! WSB is so back        0        NaN   
1  Retail giant TARGET has now declined for 10 co...      -15        NaN   
2  How do you feel about a sitting president maki...        1        NaN   
3                              Who knows more? $VMAR       30        0.0   
4  The Trump administration is begging Xi Jinping...        1        NaN   

   upvote_ratio                                               text  \
0           NaN                                   Calls on retards   
1           NaN  Stunt as in like why did they even make a big ...   
2           NaN                  Seeing lots of red in the ticker.   
3          0.98  Vision Marine Technologies Inc. is rewriting t...   
4           NaN                           He didn’t say thank you.   

   subjectivity  polarity  sentiment  \
0      1.000000 -0.900000       -1.0   
1      0.177778  0.083333        1.0   
2      0.000000  0.000000        0.0   
3      0.646970  0.216383        1.0   
4      0.000000  0.000000        0.0   

                                            entities  label  
0                                                 []   -1.0  
1              ['Stunt', 'company', 'deal', 'place']    0.0  
2                                         ['ticker']    0.0  
3  ['watercraft', 'skill', 'power', ']', 'feat', ...    1.0  
4                                                 []   -1.0

In [15]:
# check the shape
# 847 rows and 12 columns
data.shape

(847, 12)

## Data Cleaning

Let's deal with the missing values first.

In [ ]:
# check for missing values
data.isnull().sum()

type              0
subreddit         0
title             0
upvotes           0
downvotes       783
upvote_ratio    783
text              0
subjectivity      0
polarity          0
sentiment         0
entities          0
label             0
dtype: int64

Note that 783 / 847 = 0.924, imples that **92.4% of the data is missing** which is such a large percentage that it may indicate that the data is not suitable for analysis. For simply, I'll just drop them.

In [ ]:
# drop columns "downvotes" and "upvote_ratio" 
data.drop(columns = ['downvotes', 'upvote_ratio'], axis=1, inplace=True)
data.shape

C:\Users\user\AppData\Local\Temp\ipykernel_2724\1562961451.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.drop(columns = ['downvotes', 'upvote_ratio'], axis=1, inplace=True)


(847, 10)

Now let's check if there are duplicate data.

In [28]:
# check duplicate
print(f"Original data: {sum(df.duplicated())} rows are duplicated")
print(f"Processed data: {sum(data.duplicated())} rows are duplicated")

Original data: 0 rows are duplicated
Processed data: 21 rows are duplicated


We can see that there are 21 duplicated data in our processed dataset. However, I'll choose to preserve them since their `post_id`, which is dropped before, is unique for each post.

# Phase 2

In [ ]:
### Begin Assignment Here